ICE with `#[derive(Copy)]`, missing trait bound, and blanket impl

I have tested this on both 1.13 and on nightly-2016-12-28. This this is the script to reproduce: https://is.gd/PWV89w

The code is invalid due to missing constraint `Col::SqlType: IntoNullable` on line 26. The expected error is:

```
error[E0277]: the trait bound `<Col as Expression>::SqlType: NotNull` is not satisfied
  --> foo.rs:25:1
   |
25 |   pub enum ColumnInsertValue<Col, Expr> where
   |  _^ starting here...
26 | |     Col: Column,
27 | |     Expr: Expression<SqlType=<Col::SqlType as IntoNullable>::Nullable>,
28 | | {
29 | |     Expression(Col, Expr),
30 | |     Default(Col),
31 | | }
   | |_^ ...ending here: the trait `NotNull` is not implemented for `<Col as Expression>::SqlType`
   |
   = help: consider adding a `where <Col as Expression>::SqlType: NotNull` bound
   = note: required because of the requirements on the impl of `IntoNullable` for `<Col as Expression>::SqlType`
```

But the compiler instead ICEs. If `#[derive(Copy)]` is removed, the compiler gives the error message as expected. If the blanket `impl<T: NotNull> IntoNullable for T` (lines 10-12 in the script) is removed, the compiler gives the error message as expected.

Errors today, which is I think correct. E-needstest.

```
error[E0277]: the trait bound `<Col as Expression>::SqlType: NotNull` is not satisfied
  --> <anon>:16:10
   |
16 | #[derive(Copy)]
   |          ^^^^ the trait `NotNull` is not implemented for `<Col as Expression>::SqlType`
   |
   = help: consider adding a `where <Col as Expression>::SqlType: NotNull` bound
   = note: required because of the requirements on the impl of `IntoNullable` for `<Col as Expression>::SqlType`

error: aborting due to previous error

```

--- a/src/test/compile-fail/isssue-38821.rs
+++ b/src/test/compile-fail/isssue-38821.rs
@@ -1,0 +1,43 @@
+// Copyright 2017 The Rust Project Developers. See the COPYRIGHT
+// file at the top-level directory of this distribution and at
+// http://rust-lang.org/COPYRIGHT.
+//
+// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
+// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
+// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
+
+pub struct Nullable<T: NotNull>(T);
+
+pub trait NotNull {}
+
+pub trait IntoNullable {
+    type Nullable;
+}
+
+impl<T: NotNull> IntoNullable for T {
+    type Nullable = Nullable<T>;
+}
+
+impl<T: NotNull> IntoNullable for Nullable<T> {
+    type Nullable = Nullable<T>;
+}
+
+pub trait Expression {
+    type SqlType;
+}
+
+pub trait Column: Expression {}
+
+#[derive(Debug, Copy, Clone)]
+//~^ ERROR the trait bound `<Col as Expression>::SqlType: NotNull` is not satisfied
+pub enum ColumnInsertValue<Col, Expr> where
+    Col: Column,
+    Expr: Expression<SqlType=<Col::SqlType as IntoNullable>::Nullable>,
+{
+    Expression(Col, Expr),
+    Default(Col),
+}
+
+fn main() {}

